File: proposed-projects/round-trip-experiment/reverse/20260909T024849Z_b96078c6/20260910T062159Z_d79086d3/ported_implementation/gbnf/grammar_graph/get_serialized_rule_key.py
from __future__ import annotations

import json

from .type_guards import is_rule_char, is_rule_char_exclude, is_rule_end, is_rule_ref

KEY_TRANSLATION = {
    "RuleEnd": 0,
    "RuleChar": 1,
    "RuleCharExclude": 2,
}


def _serialize(value: object) -> str:
    return json.dumps(value, separators=(",", ":"))


def get_serialized_rule_key(rule: object) -> str:
    if is_rule_end(rule):
        return f"{KEY_TRANSLATION['RuleEnd']}"

    if is_rule_char(rule):
        return f"{KEY_TRANSLATION['RuleChar']}-{_serialize(rule.value)}"

    if is_rule_char_exclude(rule):
        return f"{KEY_TRANSLATION['RuleCharExclude']}-{_serialize(rule.value)}"

    if is_rule_ref(rule):
        return f"3-{rule.value}"

    raise ValueError(f"Unknown rule type: {rule}")

File: packages/gbnf-experiment/data/20260909T124157Z_4ee67ba2/ported_implementation/src/grammar-graph/get-serialized-rule-key.ts
import { RuleChar, RuleCharExclude, RuleEnd, type UnresolvedRule } from './grammar-graph-types.ts';
import { isRuleChar, isRuleCharExclude, isRuleEnd, isRuleRef } from './type-guards.ts';

export const KEY_TRANSLATION = {
  [RuleEnd.name]: 0,
  [RuleChar.name]: 1,
  [RuleCharExclude.name]: 2,
} as const;

export const getSerializedRuleKey = (rule: UnresolvedRule): string => {
  if (isRuleEnd(rule)) {
    return `${KEY_TRANSLATION[RuleEnd.name]}`;
  }

  if (isRuleChar(rule)) {
    return `${KEY_TRANSLATION[RuleChar.name]}-${JSON.stringify(rule.value)}`;
  }

  if (isRuleCharExclude(rule)) {
    return `${KEY_TRANSLATION[RuleCharExclude.name]}-${JSON.stringify(rule.value)}`;
  }

  if (isRuleRef(rule)) {
    return `3-${rule.value}`;
  }

  throw new Error(`Unknown rule type: ${rule}`);
};
